tasks/fileReplaceTask.go

```diff
@@ -0,0 +1,400 @@
+
+	rollback := false
+	backupFilename := ""
+	makeBackup := frt.BackupExtension != ""
+
+	if makeBackup {
+		backupFilename = makeBackupFilename(origFilename, frt.BackupExtension)
+	}
+
+	origFileContents, err := frte.FsManager.ReadFile(origFilename)
+	if err != nil {
+		execRes.Err = err
+		return execRes
+	}
+
+	origfileInfo, err = frte.FsManager.Stat(origFilename)
+	if err != nil {
+		execRes.Err = err
+		return execRes
+	}
+
+	var updatedFileContents string
+	var replacementCount int
+	var additionsCount int
+
+	match := frt.PatternCompiled.MatchString(origFileContents)
+	if match {
+		updatedFileContents, replacementCount = ReplaceUsingRegexpWithCount(
+			origFileContents,
+			frt.PatternCompiled,
+			frt.Repl,
+			frt.Count)
+	} else {
+		newContent := frt.Repl
+		if frt.NotFoundContent != "" {
+			newContent = frt.NotFoundContent
+		}
+		if frt.AppendIfNotFound {
+			updatedFileContents = origFileContents + newContent
+			additionsCount = 1
+		} else if frt.PrependIfNotFound {
+			updatedFileContents = newContent + origFileContents
+			additionsCount = 1
+		}
+	}
+
+	// will only be non-nil if the original contents have been updated
+	if updatedFileContents != "" {
+		if makeBackup {
+			err := frte.FsManager.WriteFile(backupFilename, origFileContents, origfileInfo.Mode())
+			if err != nil {
+				execRes.Err = err
+				return execRes
+			}
+			logrus.Debugf("created backup file %s for original file %s", backupFilename, origFilename)
+		}
+
+		err := frte.FsManager.WriteFile(frt.Name, updatedFileContents, origfileInfo.Mode())
+		if err != nil {
+			if !rollback {
+				execRes.Err = err
+				return execRes
+			}
+
+			removeBackupErr := frte.FsManager.Remove(backupFilename)
+			if removeBackupErr != nil {
+				logrus.Infof("failed to remove backup file: %v", removeBackupErr)
+			}
+			execRes.Err = err
+			return execRes
+		}
+
+		frt.Updated = true
+
+		logrus.Debugf("updated file contents for %s", origFilename)
+	}
+
+	if frt.Updated && execRes.Err == nil {
+		execRes.Comment = "File updated"
+		if replacementCount > 0 {
+			execRes.Changes["count"] = fmt.Sprintf("%d replacement(s) made", replacementCount)
+		} else if additionsCount > 0 {
+			execRes.Changes["count"] = fmt.Sprintf("%d addition(s) made", additionsCount)
+		}
+	}
+
+	execRes.Duration = time.Since(start)
+
+	logrus.Debugf("the task '%s' is finished for %v", task.GetPath(), execRes.Duration)
+	return execRes
+}
+
+func makeBackupFilename(origFilename string, ext string) (backupFilename string) {
+	return origFilename + "." + ext
+}
+
+func ReplaceUsingRegexpWithCount(contents string, re *regexp.Regexp, repl string, maxRepl int) (newContents string, replacementCount int) {
+	count := 0
+
+	replContents := re.ReplaceAllStringFunc(contents, func(matchStr string) string {
+		if maxRepl == 0 || count < maxRepl {
+			// replace again, this time using matched fragment with replacement using closure captured 're'
+			count++
+			return re.ReplaceAllString(matchStr, repl)
```

Comment:
Because the reg ex and `repl` strings might include reg ex related groups or similar. if we don't replace again, then the group substitutions don't get applied. i've added another comment line to make this clear.